### ISSUE TITLE
How to get from hand kp3d to kp2d? 

### ISSUE THREAD

**Issue by relh** [zc-alexfan/arctic#29]
Hi! I've been trying to project the 3d keypoints to 2d on my own.. 

I've been using the `mano.j3d.full.r` and `mano.j2d.norm.r`. I'm just trying to understand the full pseudocode involved in https://github.com/zc-alexfan/arctic/blob/7e3f736a95d7c461dbd93068de4367374c68a5ac/src/datasets/arctic_dataset.py.

It seems like the 3D is loaded (EDIT: just kidding I see now that they come from different sources...) and then:
1) 2D drops Z and adds a homogenous 1.0
2) 2D is scaled down by 0.3 (args.ego_image_scale)
3) A transformation matrix is built in j2d_processing that accounts for some center/scale (which is I think [420., 300.]?) 
Yet when looking at the intrinsics matrix, the center point (106.0 110.0) seems more representative of the (224, 224) resolution?

There's also a `mano.transl.r` hidden in the dataset keys, but I've only gotten things to render roughly closely by using the wrist-position in the `mano.j3d.full.r`. 

Any help elaborating what things are happening in step 3) would be really appreciated! My major issue is that I want a system that accurately recovers the 3D keypoints but to recover those and have it match the 2D keypoints I need to be able to project the 3D keypoints to 2D (with/maybe without accounting for distortion -- whatever is easiest).

**Comment by relh** [comment]
Should I be just using `mano.j3d.cam.r`? I didn't see that key in the dataset but maybe I need to change my settings..

**Comment by relh** [comment]
Somehow the existence of this function https://github.com/zc-alexfan/arctic/blob/7e3f736a95d7c461dbd93068de4367374c68a5ac/common/camera.py#L408 makes me think that there's not existing correspondence between the 3D and 2D ground truth outside of fitting a transformation per instance to allow for projection?

**Comment by relh** [comment]
Okay, the new summary of this issue is this:

I see that kp3d and kp2d are sourced from data_cam and data_2d respectively. I am curious if there is a method by which we can project kp3d to 2d and arrive at the results of kp2d. I am mostly interested in the hands. 

If there's some code snippet somewhere that I'm missing that does this I would be really interested! Any help is much appreciated.

**Comment by zc-alexfan** [comment]
Sorry for the late reply. I was hosting a HANDS workshop at ICCV during the week. 

The problem description is a bit unclear to me. However, I can describe how I get from 3D to 2D. 

Everything in [`data_cam`](https://github.com/zc-alexfan/arctic/blob/7e3f736a95d7c461dbd93068de4367374c68a5ac/src/datasets/arctic_dataset.py#L35) is in the camera coordinate of the full resolution image (2K images). Everything in `data_2d` are 2d projection on the full resolution image. 

Since the raw images are in 2K, we have to take a crop around the object before feeding to the network. Thus, any 2D supervision needs to be adjusted correspondingly (including data augmentation such as rotation, scaling and cropping). 


For 2d keypoints, it is easy. We just apply the affine transformation to the original-resolution 2D keypoints via `j2d_processing`. Note that this function also normalize 2d keypoints at the end. Any dataset key with `norm` inside are the normalized version. The ones with the same name but without `norm` is the original version. 

The same data augmentation is applied to images via `rgb_processing`. 


Now we have rgb, 2d transformed properly. We need to adjust the 3D keypoints as well. However, 3D is tricky because operation such as data augmentatino such as rotation and translation should maintain the relative distance between hands and the object in 3D. To do that, we use the base of the object by first rigid-transforming the hand-object interaction to object canonical space (see [L39-70](https://github.com/zc-alexfan/arctic/blob/7e3f736a95d7c461dbd93068de4367374c68a5ac/src/callbacks/process/process_arctic.py#L39C17-L39C37)). Then we solve the translation of the object in the cropped image view based on 2D-3D correspondance (L79). This translation is moving the hand-object interaction to the crop space (L88-95).

The 3D in this space, can then be projected into the crop images.

**Comment by zc-alexfan** [comment]
To answer your summary, as per my answer above, 3D keypoints in `data_cam` are in the original full-resolution, same as `data_2d`. So you should be able to do this via a standard perspective projection.

**Comment by relh** [comment]
Hi Alex, thanks so much.

I think the missing piece in my projection was accounting for the translation to the hand-object space that the cropping requires.

Basically doing everything above without accounting for the object meant my projection was within 1-3 pixels a lot of the time but then occasionally was off by ~20 pixels.

Hopefully if anyone else is trying to project the 3d to 2d they'll see this. If I had paid more attention to objects I'd probably have found this earlier, doh. Thanks so much! I'll close this hopefully later today once I get my projection right 😊

**Comment by zc-alexfan** [comment]
I see. It is a bit hard to help on my end because the details are not very clear to me. But let me know if others I can clarify.

**Comment by relh** [comment]
Yeah, the details are basically:

1) I was trying to just use the ArcticDataset, without the "models", etc, that are involved in the process_data callback used in `process_arctic.py`
2) This meant I didn't have any of the additional keys there

So all I had from the normal (standard?) dataset was the `mano.j2d.norm.r` and the `mano.j3d.full.r`, and I thought that I could just do the projection from the `mano.j3d.full.r` points to the 2d, but these are always off by a little bit.

So basically now I'm trying to re-write a version of `process_data` that just does what I want without needing the `out` variable that's produced by the various models, etc. A work in progress still!

**Comment by relh** [comment]
Basically something like this, but it still doesn't line up so clearly I'm doing something wrong:

```
def process_data(inputs, targets, meta_info):
    img_res = 224

    kp3d_l_cano = targets['mano.j3d.full.l'].unsqueeze(0)
    kp3d_r_cano = targets['mano.j3d.full.r'].unsqueeze(0)

    gt_kp2d_r = targets["mano.j2d.norm.r"].unsqueeze(0)  # 2D keypoints for object base
    gt_kp2d_l = targets["mano.j2d.norm.l"].unsqueeze(0)  # 2D keypoints for object base

    # unnorm 2d keypoints
    gt_kp2d_l_cano = data_utils.unormalize_kp2d(gt_kp2d_l, img_res)
    gt_kp2d_r_cano = data_utils.unormalize_kp2d(gt_kp2d_r, img_res)

    # estimate camera translation by solving 2d to 3d correspondence
    gt_transl_l = camera.estimate_translation_k(
        kp3d_l_cano,
        gt_kp2d_l_cano,
        meta_info["intrinsics"].unsqueeze(0).cpu().numpy(),
        use_all_joints=True,
        pad_2d=True,
    ).squeeze()

    # estimate camera translation by solving 2d to 3d correspondence
    gt_transl_r = camera.estimate_translation_k(
        kp3d_r_cano,
        gt_kp2d_r_cano,
        meta_info["intrinsics"].unsqueeze(0).cpu().numpy(),
        use_all_joints=True,
        pad_2d=True,
    ).squeeze()

    # move to camera coord
    gt_model_joints_r = (kp3d_r_cano + gt_transl_r[None, :])[0]
    gt_model_joints_l = (kp3d_l_cano + gt_transl_l[None, :])[0]

    targets["mano.j3d.cam.r"] = gt_model_joints_r
    targets["mano.j3d.cam.l"] = gt_model_joints_l

    return inputs, targets, meta_info
```

**Comment by relh** [comment]
Basically something like this, but it still doesn't line up so clearly I'm doing something wrong:

```
def process_data(inputs, targets, meta_info):
    img_res = 224

    kp3d_l_cano = targets['mano.j3d.full.l'].unsqueeze(0)
    kp3d_r_cano = targets['mano.j3d.full.r'].unsqueeze(0)

    gt_kp2d_r = targets["mano.j2d.norm.r"].unsqueeze(0)  # 2D keypoints for object base
    gt_kp2d_l = targets["mano.j2d.norm.l"].unsqueeze(0)  # 2D keypoints for object base

    # unnorm 2d keypoints
    gt_kp2d_l_cano = data_utils.unormalize_kp2d(gt_kp2d_l, img_res)
    gt_kp2d_r_cano = data_utils.unormalize_kp2d(gt_kp2d_r, img_res)

    # estimate camera translation by solving 2d to 3d correspondence
    gt_transl_l = camera.estimate_translation_k(
        kp3d_l_cano,
        gt_kp2d_l_cano,
        meta_info["intrinsics"].unsqueeze(0).cpu().numpy(),
        use_all_joints=True,
        pad_2d=True,
    ).squeeze()

    # estimate camera translation by solving 2d to 3d correspondence
    gt_transl_r = camera.estimate_translation_k(
        kp3d_r_cano,
        gt_kp2d_r_cano,
        meta_info["intrinsics"].unsqueeze(0).cpu().numpy(),
        use_all_joints=True,
        pad_2d=True,
    ).squeeze()

    # move to camera coord
    gt_model_joints_r = (kp3d_r_cano + gt_transl_r[None, :])[0]
    gt_model_joints_l = (kp3d_l_cano + gt_transl_l[None, :])[0]

    targets["mano.j3d.cam.r"] = gt_model_joints_r
    targets["mano.j3d.cam.l"] = gt_model_joints_l

    return inputs, targets, meta_info
```

**Comment by zc-alexfan** [comment]
hmm. It looks correct to me. Basically, you find the denormalized 2d keypoints for the two hands separately, and use the 3D keypoints in the camera coordinate to solve for the translation.

Not sure if it will help, but AFAIK, the `estimate_translation_k` function might assume the `kp3d_r_cano` to be zero for the root. Can you try to subtract the root and see? 

Further, solving this translation for the two hands separately will not guarantee the resulting 3d keypoints will preserve their relative distance to each other, because the `estimate_translation_k` is basically a monocular depth estimation problem, and it suffers from depth ambiguity. This is the reason why I use `estimate_translation_k` on the object so that it preserves the relative distances between the two hands.

**Comment by relh** [comment]
I tried to put together both hands into a single set of joints and fit that but it didn't work great.. So I took your advice and used the object (sans some of the arti_head stuff) and it's the best result so far! Seems most every pixel is now 1-2 pixels off at most. Thanks for all your help :).

If there's a reason you think I should be using the models['arti_head'] stuff let me know, but this seems to work well. For my project, I'm probably going to just filter out from training examples where the ground truth reprojection is more than a few pixels off. Thanks for making a super cool dataset~

```
def process_data(inputs, targets, meta_info, img_res=224):
    gt_kp3d_b = targets["object.kp3d.full.b"].unsqueeze(0)
    gt_kp2d_b = targets["object.kp2d.norm.b"].unsqueeze(0)
    gt_kp2d_b_cano = data_utils.unormalize_kp2d(gt_kp2d_b, img_res)

    # estimate camera translation by solving 2d to 3d correspondence
    gt_transl = camera.estimate_translation_k(
        gt_kp3d_b, #combo_kp3d,
        gt_kp2d_b_cano, #combo_kp2d,
        meta_info["intrinsics"].unsqueeze(0).cpu().numpy(),
        use_all_joints=True,
        pad_2d=True,
    ).squeeze()

    # move to camera coord
    targets["mano.j3d.cam.r"] = targets['mano.j3d.full.r'] + gt_transl[None, :]
    targets["mano.j3d.cam.l"] = targets['mano.j3d.full.l'] + gt_transl[None, :]
    return inputs, targets, meta_info
```

**Comment by relh** [comment]
I checked this some more and it still doesn't work, going to just give up on lining up the 3d keypoints or use the articulation head stuff.

**Comment by zc-alexfan** [comment]
Maybe you can share examples where it doesn't work?
